### ISSUE TITLE
full rebuild if a class contains a constant

### ISSUE THREAD

**Issue by jiqimaogou** [gradle/gradle#2767]
org.gradle.api.internal.tasks.compile.incremental.JavaChangeProcessor is full rebuild if the class contains a constant, because
`        DependentsSet actualDependents = this.previousCompilation.getDependents(className, Collections.emptySet());
`

the new constant is `Collections.emptySet()`, so the new previous constant is compare to the empty, it always true.

### Expected Behavior
if the class constant is not changed, it will not full rebuild.

### Current Behavior
the class rebuild if the class contains a constant.

**Comment by blindpirate** [comment]
Thanks for reporting, but I'm not sure I understand your question. Are you saying if a class contains a constant field, e.g.

```
class A{
    public static String CONSTANT="";
}
```

the class will always be compiled in every build?

What's your environment (OS, JDK, Gradle version etc.)? It would be great if you could provide us a code snippet to reproduce this issue.

**Comment by jiqimaogou** [comment]
yes,
```
class A{
    public static String CONSTANT="";

    public void a() {
        .... <<<<change something
    }
}
```

first, I make my java compilation incremental through `options.incremental = true`.
then If I change the method body code of a class contains a constant field, it will always full rebuild the entire source. 

My JDK is 1.8.

I debug the gradle source code [JavaChangeProcessor#line39](https://github.com/gradle/gradle/blob/v4.1.0/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/JavaChangeProcessor.java#L39)

`previousCompilation.getDependents(className, Collections.<Integer>emptySet()) ` the second paramter is `Collections.<Integer>emptySet()` which is the current class constant value, but it does not analysis the current source file to fill the Collection. It always is empty. when it compare to the previous constant value, `actualDependents.isDependencyToAll()` is always true. 

```
        if (actualDependents.isDependencyToAll()) { <<< always true
            spec.setFullRebuildCause(actualDependents.getDescription(), input.getFile()); <<< 
            return;
        }
```

**Comment by jiqimaogou** [comment]
first compile code
```
package com.androidapi23;

import android.nfc.Tag;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

public class MainActivity extends AppCompatActivity {
public static final String Tag = "mainActivity";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }
}
```

```
import static com.androidapi23.MainActivity.Tag;

/**
 * Created by chenhjohn on 2017/8/22.
 */

public class MainTest {
    public void a() {
        System.out.println(Tag);
    }
}
```

second compile
```
```
package com.androidapi23;

import android.nfc.Tag;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

public class MainActivity extends AppCompatActivity {
public static final String Tag = "mainActivity";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        system.out.println("hello") <<< add code this.
    }
}
```

run gradle build with --info. it will output the log
```
Full recompilation is required because 'MainActivity.java' was changed. Analysis took 0.014 secs.
```

**Comment by blindpirate** [comment]
I'm sorry, perhaps I misunderstand you, but if you add that line into `MainActivity.java`, it is definitely changed... what's the relationship between this change and the constant field? I mean, it's recompiled because you add the line `system.out.println("hello")`, and it has nothing to do with the constant field.

**Comment by oehme** [comment]
This is expected, because we don't analyze the source code, so we don't know that the constant didn't change. We only have finer grained detection for bytecode changes on the classpath.

Possible workarounds: 

- Make that constant private and expose a static method instead
- Put it in a class that changes less often

**Comment by blindpirate** [comment]
Cool, I think I misunderstand the incremental thing. Thanks for clarifying.

**Comment by oehme** [comment]
This is a special case for constants. Everything else can be analysed properly through the bytecode. But constant refernces aren't visible in the byte code. So we stay on the safe side and just assume that *every* other class uses that constant.

**Comment by blindpirate** [comment]
@jiqimaogou , I think oehme's explanation is pretty clear, so I'm to close this issue. Please feel free to reopen it if you still have questions.

**Comment by shwenzhang** [comment]
@oehme , I think it is not good way to rebuild all class if contain a constant. 

We can javac all the changed java classes first, then judge whether really has a constant field changed?

**Comment by oehme** [comment]
We are aware that this is not the best heuristic, but it does the job relatively well. It would be interesting to know in what kind of code base this becomes a problem. E.g. what kinds of code bases change constants all the time? Android comes to mind with its `R` classes, any others?

Note that we do have ideas on how to improve this, it just isn't the highest priority right now.

**Comment by shwenzhang** [comment]
Thanks for replay. For a large project, recompilation of javac cost more than 30s with instant run build, this due to two reason:
1.annotation; It will be solved at 4.6 #1320 
2.constant; Hundreds of people work in the project, sometime many of the classes will have constant fields. Developer may confused why building need so much time event they actually didn't change any constants.

**Comment by oehme** [comment]
> sometime many of the classes will have constant fields.

Note that this is only about *public* constants. In most cases you can avoid this by using static methods instead:

Replace this: `public static final int MAGIC = 23`
With this: `public static int magic() { return 23; }`

**Comment by imeszaros** [comment]
https://blog.gradle.org/incremental-compiler-avoidance

In the above blog post, it is explained: *In Gradle 3.4, we made that behavior much smarter, and only recompile classes which could potentially be affected by the change. In other words, if the class is changed, but the constant is not, we don’t need to recompile.*

Is it still the situation in Gradle 6.1? Our project has a class with public constants, and each time the class is changed, a full rebuild is executed (the constants does NOT change). If I change the constants to be private, incremental compilation works fine.

**Comment by stevengerdes-SW** [comment]
https://docs.gradle.org/current/userguide/java_plugin.html#sec:incremental_compile implies it must have been reverted or regressed since https://blog.gradle.org/incremental-compiler-avoidance was published. It would be nice to know if this is true and why it changed back.

**Comment by cocorossello** [comment]
> Is it still the situation in Gradle 6.1? Our project has a class with public constants, and each time the class is changed, a full rebuild is executed (the constants do NOT change). If I change the constants to be private, incremental compilation works fine.

Same problem here, gradle 6.8-rc4

**Comment by donv** [comment]
Same problem with Gradle 7.0.2.  Blank line triggers full rebuild even if the constant is not changed.

**Comment by cocorossello** [comment]
I haven't tested it yet, but I think this has been fixed in the just released 7.1-RC-1 . See https://github.com/gradle/gradle/pull/16495
